File: frontend/src/interfaces/RootTaskDisplayDTO.ts
import type { GroupDisplayDTO } from "@/interfaces/GroupDisplayDTO";
import type { UserDisplayDTO } from "@/interfaces/UserDisplayDTO";
import type { TaskStatus } from "@/enums/TaskStatus";
import type { Event } from "react-big-calendar";

export interface RootTaskDisplayDTO {
  taskId: string;

  creatorUser?: UserDisplayDTO;

  group: GroupDisplayDTO;

  name: string;

  startedAt: string;

  createdAt: string;

  doneAt: string | null;

  deadline: string;

  status: TaskStatus;

  description: string | null;

  assignees: UserDisplayDTO[];
}

export type TaskEvent = Event & {
  taskId: string;
  description: string;
  assignees: { userId: string; name?: string }[];
  doneAt: string | null;
  status: string;
};

export const toTaskEvent = (task: RootTaskDisplayDTO): TaskEvent => {
  return {
    taskId: task.taskId,
    title: task.name,
    description: task.description || "",
    status: task.status.toString() || "CREATED",
    start: new Date(task.startedAt),
    end: new Date(task.deadline),
    doneAt: task.doneAt,
    assignees: task.assignees ?? [],
  };
};

File: backend/src/main/java/com/chieftain/controllers/group/dto/RootTaskDisplayDTO.java
package com.chieftain.controllers.group.dto;

import com.chieftain.controllers.user.dto.GroupDisplayDTO;
import com.chieftain.controllers.user.dto.UserDisplayDTO;
import com.chieftain.enums.TaskStatus;
import com.chieftain.models.TaskEntity;
import com.chieftain.models.TaskStatusEntity;
import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class RootTaskDisplayDTO {

  @Nonnull private UUID taskId;

  private UserDisplayDTO creatorUser;

  @Nonnull private GroupDisplayDTO group;

  private String name;

  @Nonnull private LocalDateTime startedAt;

  @Nonnull private LocalDateTime createdAt;

  @Nullable private LocalDateTime doneAt;

  @Nonnull private LocalDateTime deadline;

  @Nonnull private TaskStatus status;

  @Nullable private String description;

  private List<UserDisplayDTO> assignees = new ArrayList<>();

  public static RootTaskDisplayDTO ofEntity(TaskEntity task) {
    RootTaskDisplayDTO dto = new RootTaskDisplayDTO();

    dto.setTaskId(task.getId());
    dto.setCreatorUser(UserDisplayDTO.ofEntity(task.getCreatorUser()));
    dto.setName(task.getName());
    dto.setStartedAt(task.getStartedAt());
    dto.setCreatedAt(task.getCreatedAt());
    dto.setDoneAt(task.getDoneAt());
    dto.setDeadline(task.getDeadline());
    dto.setStatus(task.getStatus().getStatusName());
    dto.setDescription(task.getDescription());
    dto.setGroup(GroupDisplayDTO.fromGroupEntity(task.getGroup()));
    dto.setAssignees(task.getAssignees().stream().map(UserDisplayDTO::ofEntity).toList());

    return dto;
  }
}
